lib/Support/PrettyPrinter.cpp

```diff
@@ -0,0 +1,299 @@
+// * Group indentation styles: Visual and Block, set on 'begin' tokens.
+//   "Visual" is the style in the paper, offset relative to current column.
+//   "Block" is relative to current base indentation.
+// * Break: Add "Neverbreak": acts like a break re:sizing previous range,
+//   but will never be broken.  Useful for adding content to end of line
+//   that may go over margin but should not influence layout.
+// * Begin: Add "Never" breaking style, for forcing no breaks including
+//   within nested groups.  Use sparingly.  It is an error to insert
+//   a newline (Break with spaces==kInfinity) within such a group.
+// * If leftTotal grows too large, "rebase" our datastructures by
+//   walking the tokens with pending sizes (scanStack) and adjusting
+//   them by `leftTotal - 1`.  Also reset tokenOffset while visiting.
+//   This is mostly needed due to use of tokens/groups that 'never' break
+//   which can greatly increase times between `clear()`.
+// * Optionally, minimum amount of space is granted regardless of indentation.
+//   To avoid forcing expressions against the line limit, never try to print
+//   an expression in, say, 2 columns, as this is unlikely to produce good
+//   output.
+//   (TODO)
+//
+// There are many pretty-printing implementations based on this paper,
+// and research literature is rich with functional formulations based originally
+// on this algorithm.
+//
+// Implementations of note that have interesting modifications for their
+// languages and modernization of the paper's algorithm:
+// * prettyplease / rustc_ast_pretty
+//   Pretty-printers for rust, the first being useful for rustfmt-like output.
+//   These have largely the same code and were based on one another.
+//     https://github.com/dtolnay/prettyplease
+//     https://github.com/rust-lang/rust/tree/master/compiler/rustc_ast_pretty
+//   This is closest to the paper's algorithm with modernizations,
+//   and most of the initial tweaks have also been implemented here (thanks!).
+// * swift-format: https://github.com/apple/swift-format/
+//
+// If we want fancier output or need to handle more complicated constructs,
+// both are good references for lessons and ideas.
+//
+// FWIW, at time of writing these have compatible licensing (Apache 2.0).
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Support/PrettyPrinter.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace circt {
+namespace pretty {
+
+/// Destructor, anchor.
+PrettyPrinter::Listener::~Listener() = default;
+
+/// Add token for printing.  In Oppen, this is "scan".
+void PrettyPrinter::add(Token t) {
+  // Add token to tokens, and add its index to scanStack.
+  auto addScanToken = [&](auto offset) {
+    auto right = tokenOffset + tokens.size();
+    scanStack.push_back(right);
+    tokens.push_back({t, offset});
+  };
+  llvm::TypeSwitch<Token *, void>(&t)
```

Comment:
Nit: I think `void` can be elided? Ditto below. This is more stylistic and explicit designation of the return type is not bad.

Reply:
Oh, strange, wonder why I added that.  Thanks, removed!